libs/MeshKernel/src/Mesh2DIntersections.cpp

```diff
@@ -0,0 +1,363 @@
+                                                                      const std::vector<bool>& vistedEdges) const
+{
+    UInt crossedSegmentIndex = constants::missing::uintValue;
+    UInt crossedEdgeIndex = constants::missing::uintValue;
+    bool isSeedFound = false;
+
+    // Find starting edge and segment
+    for (UInt segmentIndex = 0; segmentIndex < polyLine.size() - 1; ++segmentIndex)
+    {
+        for (UInt edgeIndex = 0; edgeIndex < mesh.GetNumEdges(); ++edgeIndex)
+        {
+            // edge already crossed, nothing to do
+            if (vistedEdges[edgeIndex])
+            {
+                continue;
+            }
+
+            const auto [isEdgeCrossed,
+                        intersectionPoint,
+                        crossProductValue,
+                        adimensionalPolylineSegmentDistance,
+                        adimensionalEdgeDistance] = AreSegmentsCrossing(polyLine[segmentIndex],
+                                                                        polyLine[segmentIndex + 1],
+                                                                        mesh.m_nodes[mesh.m_edges[edgeIndex].first],
+                                                                        mesh.m_nodes[mesh.m_edges[edgeIndex].second],
+                                                                        false,
+                                                                        mesh.m_projection);
+            if (!isEdgeCrossed)
+            {
+                continue;
+            }
+
+            crossedSegmentIndex = segmentIndex;
+            crossedEdgeIndex = edgeIndex;
+            isSeedFound = true;
+            break;
+        }
+
+        if (isSeedFound)
+        {
+            break;
+        }
+    }
+
+    return {isSeedFound, crossedEdgeIndex, crossedSegmentIndex};
+}
+
+std::tuple<bool, UInt, UInt, double, double, double> Mesh2DIntersections::GetNextEdgeIntersection(const std::vector<Point>& polyLine,
+                                                                                                  UInt edgeIndex,
+                                                                                                  UInt firstIndex,
+                                                                                                  UInt secondIndex,
+                                                                                                  int direction) const
+{
+    if (direction != 1 && direction != -1)
+    {
+        throw AlgorithmError("Invalid direction");
+    }
+
+    UInt numSteps = 0;
+    bool intersectionFound = false;
+    double crossProductValue = constants::missing::doubleValue;
+    double adimensionalPolylineSegmentDistance = constants::missing::doubleValue;
+    double adimensionalEdgeDistance = constants::missing::doubleValue;
+    const auto polyLineSize = static_cast<UInt>(polyLine.size());
+
+    const auto checkPolyLineIndex = [&]
+    {
```

Comment:
Does this compile ok? with no parentheses

Reply:
yes, it works